packages/cinterop/src/commonMain/kotlin/io/realm/kotlin/internal/interop/CoreError.kt

```diff
@@ -0,0 +1,69 @@
+package io.realm.kotlin.internal.interop
+
+/**
+ * Wrapper for C-API `realm_error_t`.
+ * See https://github.com/realm/realm-core/blob/master/src/realm.h#L231
+ */
+class CoreError(
+    categoriesNativeValue: Int,
+    val errorCodeNativeValue: Int,
+    messageNativeValue: String?,
+    // These are represent in the C-API, but not populated by Core.
+    // path: String,
+    // userError: Throwable?
+) {
+    val categories: CategoryFlags = CategoryFlags((categoriesNativeValue))
+    val errorCode: ErrorCode? = ErrorCode.of(errorCodeNativeValue)
+    val message = messageNativeValue
+
+    operator fun contains(category: ErrorCategory): Boolean = category in categories
+}
+
+data class CategoryFlags(val categoryFlags: Int) {
+
+    companion object {
+        /**
+         * See error code mapping to categories here:
+         * https://github.com/realm/realm-core/blob/master/src/realm/error_codes.cpp#L29
+         *
+         * In most cases, only 1 category is assigned, but some errors have multiple. So instead of
+         * overwhelming the user with many categories, we only select the most important to show
+         * in the error message. "important" is of course tricky to define, but generally
+         * we consider vague categories like [ErrorCategory.RLM_ERR_CAT_RUNTIME] as less important
+         * than more specific ones like [ErrorCategory.RLM_ERR_CAT_JSON_ERROR].
+         *
+         * In the current implementation, categories between index 0 and 7 are considered equal
+         * and the order is somewhat arbitrary. No error codes has multiple of these categories
+         * associated either.
+         */
+        val CATEGORY_ORDER: List<ErrorCategory> = listOf(
+            ErrorCategory.RLM_ERR_CAT_CUSTOM_ERROR,
+            ErrorCategory.RLM_ERR_CAT_WEBSOCKET_ERROR,
+            ErrorCategory.RLM_ERR_CAT_SYNC_ERROR,
+            ErrorCategory.RLM_ERR_CAT_SERVICE_ERROR,
+            ErrorCategory.RLM_ERR_CAT_JSON_ERROR,
+            ErrorCategory.RLM_ERR_CAT_CLIENT_ERROR,
+            ErrorCategory.RLM_ERR_CAT_SYSTEM_ERROR,
+            ErrorCategory.RLM_ERR_CAT_FILE_ACCESS,
+            ErrorCategory.RLM_ERR_CAT_HTTP_ERROR,
+            ErrorCategory.RLM_ERR_CAT_INVALID_ARG,
+            ErrorCategory.RLM_ERR_CAT_APP_ERROR,
+            ErrorCategory.RLM_ERR_CAT_LOGIC,
+            ErrorCategory.RLM_ERR_CAT_RUNTIME,
+        )
+    }
+
+    /**
+     * Returns a description of the most important category defined in [categoryFlags].
+     * If no known categories are found, the integer values for all the categories is returned
+     * as debugging information.
+     */
+    val description: String = CATEGORY_ORDER.firstOrNull { category ->
+        category in this
```

Comment:
I would probably also have gone for this if it was me writing it, but if feels a bit magic and would maybe be easier to review with an explicit `contains` call.